src/platform/windows/display_wgc.cpp

```diff
@@ -0,0 +1,290 @@
+        return -1;
+      }
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: "sv << e.code();
+      return -1;
+    }
+
+    DXGI_OUTPUT_DESC output_desc;
+    uwp_device = d3d_comhandle.as<winrt::IDirect3DDevice>();
+    display->output->GetDesc(&output_desc);
+
+    auto monitor_factory = winrt::get_activation_factory<winrt::GraphicsCaptureItem, IGraphicsCaptureItemInterop>();
+    if (monitor_factory == nullptr ||
+        FAILED(status = monitor_factory->CreateForMonitor(output_desc.Monitor, winrt::guid_of<winrt::IGraphicsCaptureItem>(), winrt::put_abi(item)))) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: [0x"sv << util::hex(status).to_string_view() << ']';
+      return -1;
+    }
+
+    if (config.dynamicRange)
+      display->capture_format = DXGI_FORMAT_R16G16B16A16_FLOAT;
+    else
+      display->capture_format = DXGI_FORMAT_B8G8R8A8_UNORM;
+
+    try {
+      frame_pool = winrt::Direct3D11CaptureFramePool::CreateFreeThreaded(uwp_device, static_cast<winrt::Windows::Graphics::DirectX::DirectXPixelFormat>(display->capture_format), 2, item.Size());
+      capture_session = frame_pool.CreateCaptureSession(item);
+      frame_pool.FrameArrived({ this, &wgc_capture_t::on_frame_arrived });
+      capture_session.IsBorderRequired(false);
+      capture_session.StartCapture();
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: "sv << e.code();
+      return -1;
+    }
+    return 0;
+  }
+
+  void
+  wgc_capture_t::on_frame_arrived(winrt::Direct3D11CaptureFramePool const &sender, winrt::IInspectable const &) {
+    // this PRODUCER runs in a separate thread spawned by the frame pool. to
+    // retain parity with the original interface, the frame will be consumed by
+    // the capture thread, not this one.
+
+    winrt::Windows::Graphics::Capture::Direct3D11CaptureFrame frame { nullptr };
+    try {
+      frame = sender.TryGetNextFrame();
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(warning) << "Failed to capture frame: "sv << e.code();
+      return;
+    }
+    if (frame != nullptr) {
+      AcquireSRWLockExclusive(&frame_lock);
+      if (produced_frame) {
+        produced_frame.Close();
+      }
+      produced_frame = frame;
+      WakeConditionVariable(&frame_present_cv);
+      ReleaseSRWLockExclusive(&frame_lock);
+    }
+  }
+
+  capture_e
+  wgc_capture_t::next_frame(std::chrono::milliseconds timeout, ID3D11Texture2D **out, uint64_t &out_time) {
+    // this CONSUMER runs in the capture thread
+    release_frame();
+
+    AcquireSRWLockExclusive(&frame_lock);
+    if (timeout.count() > 0 && SleepConditionVariableSRW(&frame_present_cv, &frame_lock, timeout.count(), 0) == FALSE) {

```

Comment:
This needs to first check if a frame is available before sleeping on the condition variable. Unlike an event, a condition variable doesn't have stored signalled/not-signalled state. It only wakes waiters that are asleep at the time `WakeConditionVariable()` is called. This means you always want to have it gated by a check that happens under your lock (see [Microsoft's sample code](https://learn.microsoft.com/en-us/windows/win32/sync/condition-variables))

Likewise, on the producer side, you don't have to call `WakeConditionVariable()` when `produced_frame` is already non-null, because you're not making a state transition where the consumer could possibily be waiting (the predicate is _already_ true).

This is probably what's killing your performance. If the consumer isn't waiting at the time the producer gets a frame (N), that frame is going to be completely lost because the consumer is going to go to sleep on the CV even though there's already a frame ready for it. Only once asleep will the consumer get properly woken up with the _next_ new frame (N+1).

Reply:
Thank you for this.
To make a long story short, today I learned that condition variables don't actually store signaled state. With your suggested changes, I no longer see the performance degradation observed earlier! I wonder if the stuttering @psyke83 observed is still present.